File: dynamic_listings/python/default/AI/a_star.py
# We bootstrap the variables
openSet: list[Node] = []
closedSet: list[Node] = []
currentNode: Node = start
closedSet.append(currentNode)
while (currentNode != end):
    for n in currentNode.getAdjacentList():
        if n in closedSet:
            continue
        if n in openSet:  # Check if this path is better
            new_g: float = getPathCost(n, start)
            if new_g < n.g:
                # We found a better path from start to currentNode
                n.parent = currentNode
                n.g = new_g
                n.f = n.g + n.h
        else:
            n.parent = currentNode
            n.g = getPathCost(n, start)
            n.h = getHeuristicCost(n, end)
            n.f = n.g + n.h
            openSet.append(currentNode)

    if not openSet:
        # We exhausted all possibilities
        break
    openSet = sorted(openSet, key=lambda n: n.f)
    # Since openset is ordered by f, the first element is the one with the lowest total cost
    currentNode = openSet[0]
    openSet.remove(currentNode)
    closedSet.append(currentNode)

# Reconstruct the path like in the greedy "best-first" algorithm

File: dynamic_listings/rust/default/AI/a_star.rs
let mut open_set = Vec::new();
let mut closed_set = Vec::new();
let mut current_node = start.clone();

closed_set.push(current_node.clone());

while current_node != end {
    for mut n in current_node.adjacent_list() {
        // closed_set Contains N
        if let Some(pos) = closed_set.iter().position(|x| x == &current_node) && pos != closed_set.len() - 1 {
            // We already analyzed this node, skip it
            continue;
        } else {
            if let Some(pos) = open_set.iter().position(|x| x == &current_node) && pos != open_set.len() - 1 {
                let new_g = path_cost(&n, &start);
                if new_g < n.g {
                    // We found a better path from start to currentNode
                    n.parent = Some(Box::new(current_node.clone()));
                    n.g = new_g;
                    n.f = n.g + n.h;
                }
            } else {
                n.parent = Some(Box::new(current_node.clone()));
                n.g = path_cost(&n, &start);
                n.h = heuristic_cost(&n, &end);
                n.f = n.g + n.h;
                open_set.push(current_node.clone());
            }
        }

        // Order openSet by f
        open_set.sort_by(|a, b| a.f.total_cmp(&b.f));
        // Since openset is ordered by f, the first element is the one with the lowest total cost
        if let Some(first) = open_set.pop() {
            current_node = first;
        } else {
            // We exhausted all the possibilities
            break;
        }
        closed_set.push(current_node.clone());
    }
}
